components/datetime/src/provider/packed_pattern.rs

```diff
@@ -0,0 +1,480 @@
+            .iter()
+            .zip(fallbacks.iter())
+            .zip(chunks.iter_mut())
+        {
+            if let Some(pattern) = pattern {
+                if pattern != fallback {
+                    *chunk = match elements.iter().position(|p| p == pattern) {
+                        Some(i) => i + 1,
+                        None => {
+                            elements.push(pattern);
+                            elements.len()
+                        }
+                    }
+                }
+            }
+        }
+
+        // Check to see if we need to switch to Q=1 mode
+        #[allow(clippy::unwrap_used)] // the array is nonempty
+        if chunks.iter().max().unwrap() >= &0x8 {
+            // one pattern per table cell
+            header |= 0x4;
+            elements.truncate(s_offset + 1);
+            elements.extend(
+                variant_patterns
+                    .into_iter()
+                    .zip(fallbacks.iter())
+                    .map(|(pattern, fallback)| pattern.unwrap_or(fallback)),
+            );
+        } else {
+            // per-cell offsets
+            let mut shift = 3;
+            for chunk in chunks.iter() {
+                header |= chunk << shift;
+                shift += 3;
+            }
+        }
+
+        // Now we can build the data representation
+        let elements = elements
+            .iter()
+            .map(|plural_elements| {
+                plural_elements
+                    .as_ref()
+                    .map(|pattern| (pattern.metadata.to_four_bit_metadata(), &*pattern.items))
+            })
+            .collect::<Vec<_>>();
+        PackedPatternsV1 {
+            #[allow(clippy::unwrap_used)] // the header fits in 21 bits
+            header: u32::try_from(header).unwrap(),
+            elements: VarZeroVec::from(&elements),
+        }
+    }
+
+    fn simplify(&mut self) {
+        if self.variant0.as_ref() == Some(&self.standard) {
+            self.variant0 = None;
+        }
+        if self.variant1.as_ref() == Some(&self.standard) {
+            self.variant1 = None;
+        }
+    }
+}
+
+pub(crate) enum PackedSkeletonVariant {
+    Standard,
+    Variant0,
+    Variant1,
+}
+
+impl PackedPatternsV1<'_> {
+    pub(crate) fn get(
+        &self,
+        length: NeoSkeletonLength,
+        variant: PackedSkeletonVariant,
+    ) -> PatternBorrowed {
+        use NeoSkeletonLength::*;
+        use PackedSkeletonVariant::*;
+        let lms = self.header & 0x3;
+        let pattern_index = if matches!(variant, Standard) {
+            // Standard pattern (first column)
+            match (length, lms) {
+                (Long, _) => 0,
+                (Medium, 0 | 2) => 0,
+                (Medium, _) => 1,
+                (Short, 0) => 0,
+                (Short, 1 | 2) => 1,
+                (Short, _) => 2,
+            }
+        } else {
+            let s_offset = match lms {
+                0 => 0,
+                1 | 2 => 1,
```

Comment:
nit: here and above, when we mention different LMS values by number we should document them as `LM,S` etc. 

Alternatively, we could have an `enum LMS { L_M_S, LM_S, ...}` (yes i'm okay with underscores for internal enums) with a function that extracts it off of a header. Or a set of constants.

Reply:
Done in 6b30bd88b04128dcb83469a43bd598d1b9640e41